src/gp_dag.cpp

```diff
@@ -106,6 +109,87 @@ EigenVectorXd GPDAG::BuildUniformQ() const {
   return q;
 }
 
+Node::NodePtrVec GPDAG::GenerateAllTrees() const {
+  std::vector<Node::NodePtrVec> trees_below(NodeCount());
+
+  auto GetSubtrees = [&trees_below](const GPDAGNode *node,
+                                          Node::NodePtrVec &rotated_subtrees,
+                                          Node::NodePtrVec &ordered_subtrees) {
+    for (const bool rotated : {false, true}) {
+      for (const auto &child_id : node->GetLeafward(rotated)) {
+        for (const auto &sub_tree : trees_below.at(child_id)) {
+          rotated ? rotated_subtrees.push_back(sub_tree) :
+          ordered_subtrees.push_back(sub_tree);
+        }
+      }
+    }
+  };
+
+  auto MergeTrees = [](size_t node_id,
+                       Node::NodePtrVec &trees,
+                       Node::NodePtrVec &rotated_subtrees,
+                       Node::NodePtrVec &ordered_subtrees) {
+    for (size_t i = 0; i < rotated_subtrees.size(); i++) {
+      auto &rotated_subtree = rotated_subtrees.at(i);
+      for (size_t j = 0; j < ordered_subtrees.size(); j++) {
+        auto &ordered_subtree = ordered_subtrees.at(j);
+        Node::NodePtr new_tree = Node::Join(ordered_subtree, rotated_subtree, node_id);
+        trees.push_back(new_tree);
+      }
+    }
+  };
+
+  for (const auto &node_id : RootwardPassTraversal()) {
+    const auto &node = GetDagNode(node_id);
+    if (node->IsLeaf()) {
+      trees_below.at(node_id).push_back(Node::Leaf(node_id));
+    } else {
+      Node::NodePtrVec rotated_nodes, ordered_nodes;
+      GetSubtrees(node, rotated_nodes, ordered_nodes);
```

Comment:
Could you please refactor these functions so that they return vectors rather than use output arguments? I realize that we do frequently use this pattern when we need to modify something, but in general returning outputs is preferred.

Reply:
OK. Would you like to keep the lambdas or should I factor them out as their own functions within the class?

Reply:
I would prefer factoring out to functions. Thanks.

Reply:
I'll keep the lambdas but have changed the functions to return vectors.